python/triton/tools/ttc.py

```diff
@@ -0,0 +1,237 @@
+    jitted: Dict[str, ASTGeneratingObject],
+) -> AOTMetadataWithAST:
+    """
+
+    Take metadata and jitted functions found and make sure those match.
+    Kernels with missing jit function or metadata are skipped
+
+    :return:
+        Dict with function names (major names)
+    """
+
+    # TODO: add some checks to make sure signatures are the same
+    valid_meta_func_paris = {}
+    for conf_func in aot_meta:
+        if conf_func not in jitted:
+            print(
+                f"[Skipping] {conf_func} -- is not a valid function in sources provided"
+            )
+            continue
+        metas, fn = aot_meta[conf_func], jitted[conf_func]
+        valid_meta_func_paris[conf_func] = (metas, fn)
+
+    # check what functoin are imported and have no metadata specified
+    for fn_name in set(jitted.keys()) - set(aot_meta.keys()):
+        print(f"[Missing Meta]: {fn_name} is imported but has no metadata")
+        # TODO: try and infer the missing funcitons?
+
+    if not len(valid_meta_func_paris):
+        raise ValueError(f"Config file has no configs for source kernels")
+
+    return valid_meta_func_paris
+
+ASMObject = Dict[str, Union[str, bytes]]
+""" The `asm` attr from `CompiledKernel` """
+
+
+def compile_func(meta: AOTKernelMetadata, fn: ASTGeneratingObject) -> Tuple[ASMObject, CompileMetadata]:
+    # NOTE: we ovveride the function name here. in case we have several variants in the kerenel
+    #       (those will have same signature in C)
+    old_name = fn.__name__
+    fn.__name__ = meta.name
+
+    compile_metadata = make_compile_metadata(meta)
+    compiled = triton.compile(fn, **compile_metadata.kwargs)
+    asm = compiled.asm
+
+    fn.__name__ = old_name
+    return asm, compile_metadata
+
+
+def generate_c_code(meta_and_ast: AOTMetadataWithAST, outdir: str):
+    """
+    Code generation goes like this:
+        - each jitted function gets replicated according to the number of constexpr variants it has
+        - each variant has its own header and source (with binary/ptx stored in source)
+        - common header is generated for includes and utils
+    """
+    # C Code generation.
+    codegen = CodeGenerator()
+
+    for fn_name, (metas, jit_fn) in meta_and_ast.items():
+        for variant_idx, meta in enumerate(metas):
+            asm, comple_meta = compile_func(meta, jit_fn)
+            docstr = str(meta)
+            if jit_fn.__doc__ is not None:
+                docstr = f"{docstr}\n\t{jit_fn.__doc__}"
+            codegen.gen_kernel(kernel_meta=meta, compile_meta=comple_meta, bin_=asm['cubin'], docstring=docstr)
+
+    codegen.dump(out_dir=outdir)
+
+
+if __name__ == "__main__":
+    from pathlib import Path
+
+
+    parser = ArgumentParser(description="Triton Ahead of Time Compilation (AoT)")
+    parser.add_argument(
+        "paths",
+        nargs="+",
+        help="Path to Python source that contains JITFunction in scope (note the source will be executed)",
```

Comment:
I think we should probably aim to use an AST visitor rather than executing a source

Reply:
Switched metadata generation to AST analysis. For now supprts global constants and @triton.jit functions